### PR TITLE
pkg/client: Go module

### PR DESCRIPTION
**Pull request by williamlord**
Resolves https://github.com/prometheus-operator/prometheus-operator/issues/3269.

It's the same for us, we don't want all these transitive dependencies when we only need the client and apis packages.

I'm not a Go expert so appreciate a thorough review.

### CODE REVIEW

**Review by simonpasquier — COMMENTED** [review]
Please update the Makefile to add the extracted packages [here](https://github.com/prometheus-operator/prometheus-operator/blob/7edea80200d9f0eff58154839f1a581093d09b0d/Makefile#L64). I notice too that we don't run `go mod tidy` on `pkg/apis/monitoring`. Please update the Makefile's tidy target for this as well as for `pkg/client`.

If this PR merges, I suspect there will be issues for projects that want to depend on the `master` branch of prometheus-operator until we have `v0.44.0`. Maybe we want to do it as part of cutting v0.44.0? I'd like also to hear opinions from other maintainers about this addition in general. cc @prometheus-operator/prometheus-operator-reviewers

**Comment by paulfantom** [comment]
For me, it starts to get messy. Maybe we should consider separating `pkg/client` and `pkg/apis` into one separate repository? Or maybe group them together in one dir and tag only this? Just throwing ideas around ;)

**Comment by simonpasquier** [comment]
>  separating pkg/client and pkg/apis into one separate repository

We tried this in prometheus with tsdb and rolled back to have everything in one repository because development workflow was too tedious.

At the same time, `pkg/client` is fully generated so the public copy could leave in a separate repo (e.g. `github.com/prometheus-operator/client`) while the project would keep using the local `pkg/client` directory. The copy could be synced whenever a new release of prometheus-operator is cut?

>  group them together in one dir and tag only this?

This would add `k8s.io/client` as a dependency for folks that don't need the client bits.

The third option is to encourage people to generate the client code directly in their project if they don't want to depend on the whole prometheus-operator repository.

**Comment by williamlord** [comment]
I've updated the PR.

What does generating the client code look like? `make k8s-gen` for client target with my own package for output package?

**Comment by lilic** [comment]
Agreed about not moving these pkgs to a separate repo, as besides what was outlined, it would also be a mess to fix import paths for end users, these are our APIs so I assume we have to keep some stability we would still need to keep them in our repo for some time. 

> If this PR merges, I suspect there will be issues for projects that want to depend on the master branch of prometheus-operator until we have v0.44.0. Maybe we want to do it as part of cutting v0.44.0?

I guess we missed the 0.44 deadline here. But @simonpasquier can you explain the issues that projects would face? Thanks!

**Comment by simonpasquier** [comment]
> I guess we missed the 0.44 deadline here. But @simonpasquier can you explain the issues that projects would face? Thanks!

I suspect that we'll hit issues like https://github.com/prometheus-operator/prometheus-operator/issues/3530 because Go wouldn't find any tagged version for `github.com/prometheus-operator/prometheus-operator/pkg/client`. But external projects depending on a valid tag (e.g. not `master`) shouldn't be impacted.
Personally I'm not opposed to this patch. It adds a bit of overhead when doing a release (e.g. we need to create one more tag) but I think it's acceptable. Having a separate repo as `github.com/prometheus-operator/client` requires some work too. @paulfantom what do you think after running v0.44.0?

> What does generating the client code look like? make k8s-gen for client target with my own package for output package?

Yes, it would mean looking at the command executed by `make k8s-gen` and change the output package path. But I don't think it's viable in the long run for users, we need to fill the gap somehow.

**Comment by lilic** [comment]
I agree, I am not opposed to this either, unless Pawel thinks it put a lot more effort and complexity into the release have not gone through the new process yet myself.

**Comment by paulfantom** [comment]
After doing v0.44.0 release I don't think this would overwhelm our release process. Especially that [RELEASE.md](https://github.com/prometheus-operator/prometheus-operator/blob/master/RELEASE.md) document is written in such detail that we are almost in a copy-paste situation when doing a release. Adding another tag would add one more entry for each release on [tags page](https://github.com/prometheus-operator/prometheus-operator/tags) and pollute it a bit but I think we can live with that. :)

That said, I think we reached a consensus so let's merge it. But first @williamlord can you rebase the PR?

**Review by paulfantom — APPROVED** [review]
LGTM

**Comment by williamlord** [comment]
Rebased now and updated with v0.44.0.

File: RELEASE.md
Comment (simonpasquier) [inline RELEASE.md:54]: we need also to bump `github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring` in `pkg/client/go.mod`?

**Review by simonpasquier — APPROVED** [review]
LGTM @lilic feel free to merge if you're ok with the change.

**Comment by lilic** [comment]
Sorry it seems we have a conflict, @williamlord do you mind resolving it after that I will merge, if you by any chance can also squash some commits would be great as well, thanks for your patience and the contribution!  🎉

**Comment by williamlord** [comment]
Rebased on v0.44.1 and squashed all the commits now.

**Review by lilic — APPROVED** [review]
Thanks!

lgtm

**Comment by williamlord** [comment]
Unfortunately now the pkg/apis/monitoring package depends on k8s.io/apiextensions-apiserver which itself has a lot of dependencies.
https://github.com/prometheus-operator/prometheus-operator/pull/3703/files#diff-bdf31a6a6c5bd032faf1aaeeb90cebdf5bae06f6301372d6f51a69879976c582R7

**Comment by lilic** [comment]
@williamlord please open a new issue as we tend to miss out comments on merged PRs, thanks!

**Pull request merged**
